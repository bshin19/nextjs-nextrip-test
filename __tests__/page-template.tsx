import PageTemplate, { fetcher } from "../components/page-template";
import { fireEvent, render } from "@testing-library/react";
import { SWRConfig } from "swr";

describe("PageTemplate", () => {
  let useRouter: any, windowFetchSpy;

  const fallback = {
    "https://svc.metrotransittest.org/nextripv2/routes": [
      { route_id: "901", agency_id: 0, route_label: "METRO Blue Line" },
    ],
  };

  const directions = [
    {
      direction_id: 0,
      direction_name: "Northbound",
    },
    {
      direction_id: 1,
      direction_name: "Southbound",
    },
  ];

  beforeEach(
    () => (useRouter = jest.spyOn(require("next/router"), "useRouter"))
  );

  afterEach(() => jest.restoreAllMocks());

  it("should match failed request path", () => {
    let swrSpy = jest.spyOn(require("swr"), "default");
    swrSpy.mockReturnValue({ error: true });

    const { container } = render(<PageTemplate />);

    expect(container).toMatchSnapshot();
  });

  it("should match loading", () => {
    useRouter.mockImplementation(() => ({
      query: {},
      asPath: "https://svc.metrotransittest.org/nextripv2/routes",
    }));

    const { container } = render(
      <SWRConfig
        value={{
          fallback: {
            "https://svc.metrotransittest.org/nextripv2/routes": false,
          },
        }}
      >
        <PageTemplate />
      </SWRConfig>
    );

    expect(container).toMatchSnapshot();
  });

  describe("successful request", () => {

    it("should match", () => {
      useRouter.mockImplementation(() => ({
        query: {},
        asPath: "https://svc.metrotransittest.org/nextripv2/routes",
      }));

      const { container } = render(<PageTemplate />);

      expect(container).toMatchSnapshot();
    });

    it('should render with empty routes array', () => {
      let swrSpy = jest.spyOn(require("swr"), "default");
      swrSpy.mockReturnValue({ data: { routes: [] }});

      useRouter.mockImplementation(() => ({
        query: {},
        asPath: "https://svc.metrotransittest.org/nextripv2/routes",
      }));

      const { container } = render(<PageTemplate />);

      expect(container).toMatchSnapshot();
    })

    it("should match with all props", () => {
      useRouter.mockImplementation(() => ({
        query: {},
        asPath: "https://svc.metrotransittest.org/nextripv2/routes",
      }));

      const { container } = render(
        <SWRConfig value={{ fallback }}>
          <PageTemplate directions={directions}>
            <div>My child</div>
          </PageTemplate>
        </SWRConfig>
      );

      expect(container).toMatchSnapshot();
    });

    it('should invoke onRouteChange with the expected args', () => {
      
      const onRouteChangeListener = jest.fn()
      useRouter.mockImplementation(() => ({
        query: {},
        push: onRouteChangeListener,
        asPath: "https://svc.metrotransittest.org/nextripv2/routes",
      }));

      const { getAllByRole } = render(
        <SWRConfig value={{ fallback }}>
          <PageTemplate directions={directions}/>
        </SWRConfig>
      );

      expect(onRouteChangeListener).not.toHaveBeenCalled()

      fireEvent.change(getAllByRole('combobox')[0])

      expect(onRouteChangeListener).toHaveBeenCalledWith('/route/')

    })

    it('should invoke onDirectionChange with the expected args', () => {
      
      const onDirectionChange = jest.fn()
      const route = 'wow'
      useRouter.mockImplementation(() => ({
        query: { route },
        push: onDirectionChange,
        asPath: "https://svc.metrotransittest.org/nextripv2/routes",
      }));

      const { getAllByRole } = render(
        <SWRConfig value={{ fallback }}>
          <PageTemplate directions={directions}/>
        </SWRConfig>
      );

      expect(onDirectionChange).not.toHaveBeenCalled()

      fireEvent.change(getAllByRole('combobox')[1])

      expect(onDirectionChange).toHaveBeenCalledWith(`/route/${route}/`)

    })

    it("should match with route data", () => {
      useRouter.mockImplementation(() => ({
        query: { route: "901", direction: 0 },
        asPath: "https://svc.metrotransittest.org/nextripv2/routes",
      }));

      const { container } = render(
        <SWRConfig
          value={{
            fallback: {
              "https://svc.metrotransittest.org/nextripv2/routes": [
                {
                  route_id: "901",
                  agency_id: 0,
                  route_label: "METRO Blue Line",
                },
                {
                  route_id: "902",
                  agency_id: 0,
                  route_label: "METRO Green Line",
                },
                {
                  route_id: "906",
                  agency_id: 10,
                  route_label: "Airport Shuttle",
                },
                {
                  route_id: "903",
                  agency_id: 0,
                  route_label: "METRO Red Line",
                },
              ],
            },
          }}
        >
          <PageTemplate />
        </SWRConfig>
      );

      expect(container).toMatchSnapshot();
    });
  });
});

// just covering the code that exists here, given more time would expand on functionality and what codes we're looking for
describe("fetcher", () => {
  it("should fetch and parse json", async () => {
    expect(
      await fetcher("https://svc.metrotransittest.org/nextripv2/routes")
    ).toMatchSnapshot();
  });
});
