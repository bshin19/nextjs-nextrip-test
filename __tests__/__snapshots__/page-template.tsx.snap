// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageTemplate should match failed request path 1`] = `
<div>
  <div>
    failed to load
  </div>
</div>
`;

exports[`PageTemplate should match loading 1`] = `
<div>
  <main
    class="mx-4 my-4 p-2 pb-5 grow grid bg-base rel rounded"
  >
    <h1
      class="fs-2 text-center fg-primary leading-none mb-1 cols-12"
    >
      NexTrip: Metro Transit
    </h1>
    <h2
      class="fs-1.5 text-center mb-2 fg-secondary leading-none border-b cols-12 mx-auto"
    >
      Explore Minnesota
      <i
        class="block ml-.25 italic fs-1 fg-accent-primary"
      >
        Without Worrying About the Traffic
      </i>
    </h2>
    <div
      class="cols-12 flex flex-col mx-auto"
    >
      <label
        class="mb-.5 bold text-center"
        for="routes"
      >
        Choose a bus route:
      </label>
      <select
        class="shimmer"
        id="routes"
        name="routes"
      />
      <label
        class="mb-.5 bold text-center"
        for="directions"
      >
        Choose a direction:
      </label>
      <select
        class="shimmer"
        id="directions"
        name="directions"
      />
    </div>
    <a
      aria-label="Learn more about MetroTransit NexTrip"
      class="pointer abs t-.25 r-.75 h-2 w-4 shimmer"
      href="https://www.metrotransit.org/nextrip"
    />
  </main>
</div>
`;

exports[`PageTemplate successful request should match 1`] = `
<div>
  <main
    class="mx-4 my-4 p-2 pb-5 grow grid bg-base rel rounded"
  >
    <h1
      class="fs-2 text-center fg-primary leading-none mb-1 cols-12"
    >
      NexTrip: Metro Transit
    </h1>
    <h2
      class="fs-1.5 text-center mb-2 fg-secondary leading-none border-b cols-12 mx-auto"
    >
      Explore Minnesota
      <i
        class="block ml-.25 italic fs-1 fg-accent-primary"
      >
        Without Worrying About the Traffic
      </i>
    </h2>
    <div
      class="cols-12 flex flex-col mx-auto"
    >
      <label
        class="mb-.5 bold text-center"
        for="routes"
      >
        Choose a bus route:
      </label>
      <select
        class="shimmer"
        id="routes"
        name="routes"
      />
      <label
        class="mb-.5 bold text-center"
        for="directions"
      >
        Choose a direction:
      </label>
      <select
        class="shimmer"
        id="directions"
        name="directions"
      />
    </div>
    <a
      aria-label="Learn more about MetroTransit NexTrip"
      class="pointer abs t-.25 r-.75 h-2 w-4 shimmer"
      href="https://www.metrotransit.org/nextrip"
    />
  </main>
</div>
`;

exports[`PageTemplate successful request should match with all props 1`] = `
<div>
  <main
    class="mx-4 my-4 p-2 pb-5 grow grid bg-base rel rounded"
  >
    <h1
      class="fs-2 text-center fg-primary leading-none mb-1 cols-12"
    >
      NexTrip: Metro Transit
    </h1>
    <h2
      class="fs-1.5 text-center mb-2 fg-secondary leading-none border-b cols-12 mx-auto"
    >
      Explore Minnesota
      <i
        class="block ml-.25 italic fs-1 fg-accent-primary"
      >
        Without Worrying About the Traffic
      </i>
    </h2>
    <div
      class="cols-12 flex flex-col mx-auto"
    >
      <label
        class="mb-.5 bold text-center"
        for="routes"
      >
        Choose a bus route:
      </label>
      <select
        class="fs-1 mb-1"
        id="routes"
        name="routes"
      >
        <option
          value=""
        >
          Select an available route
        </option>
        <option
          value="901"
        >
          METRO Blue Line
        </option>
      </select>
      <label
        class="mb-.5 bold text-center"
        for="directions"
      >
        Choose a direction:
      </label>
      <select
        class="fs-1 mb-1"
        id="directions"
        name="directions"
      >
        <option
          value=""
        >
          Select a direction
        </option>
        <option
          value="0"
        >
          Northbound
        </option>
        <option
          value="1"
        >
          Southbound
        </option>
      </select>
    </div>
    <div
      class="cols-12"
    >
      <div>
        My child
      </div>
    </div>
    <a
      aria-label="Learn more about MetroTransit NexTrip"
      class="pointer abs t-.25 r-.75"
      href="https://www.metrotransit.org/nextrip"
    >
      <span
        style="box-sizing: border-box; display: inline-block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: relative; max-width: 100%;"
      >
        <span
          style="box-sizing: border-box; display: block; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; max-width: 100%;"
        >
          <img
            alt=""
            aria-hidden="true"
            src="data:image/svg+xml,%3csvg%20xmlns=%27http://www.w3.org/2000/svg%27%20version=%271.1%27%20width=%2764%27%20height=%2732%27/%3e"
            style="display: block; max-width: 100%; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px;"
          />
        </span>
        <img
          alt="MN Logo"
          data-nimg="intrinsic"
          decoding="async"
          src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
          style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
        />
        <noscript />
      </span>
    </a>
  </main>
</div>
`;

exports[`PageTemplate successful request should match with route data 1`] = `
<div>
  <main
    class="mx-4 my-4 p-2 pb-5 grow grid bg-base rel rounded"
  >
    <h1
      class="fs-2 text-center fg-primary leading-none mb-1 cols-12"
    >
      NexTrip: Metro Transit
    </h1>
    <h2
      class="fs-1.5 text-center mb-2 fg-secondary leading-none border-b cols-12 mx-auto"
    >
      Explore Minnesota
      <i
        class="block ml-.25 italic fs-1 fg-accent-primary"
      >
        Without Worrying About the Traffic
      </i>
    </h2>
    <div
      class="cols-12 flex flex-col mx-auto"
    >
      <label
        class="mb-.5 bold text-center"
        for="routes"
      >
        Choose a bus route:
      </label>
      <select
        class="fs-1 mb-1"
        id="routes"
        name="routes"
      >
        <option
          value=""
        >
          Select an available route
        </option>
        <option
          selected=""
          value="901"
        >
          METRO Blue Line
        </option>
        <option
          value="902"
        >
          METRO Green Line
        </option>
        <option
          value="906"
        >
          Airport Shuttle
        </option>
        <option
          value="903"
        >
          METRO Red Line
        </option>
      </select>
      <label
        class="mb-.5 bold text-center"
        for="directions"
      >
        Choose a direction:
      </label>
      <select
        class="fs-1 mb-1"
        disabled=""
        id="directions"
        name="directions"
      >
        <option
          value=""
        >
          Select a direction
        </option>
      </select>
    </div>
    <div
      class="cols-12"
    />
    <a
      aria-label="Learn more about MetroTransit NexTrip"
      class="pointer abs t-.25 r-.75"
      href="https://www.metrotransit.org/nextrip"
    >
      <span
        style="box-sizing: border-box; display: inline-block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: relative; max-width: 100%;"
      >
        <span
          style="box-sizing: border-box; display: block; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; max-width: 100%;"
        >
          <img
            alt=""
            aria-hidden="true"
            src="data:image/svg+xml,%3csvg%20xmlns=%27http://www.w3.org/2000/svg%27%20version=%271.1%27%20width=%2764%27%20height=%2732%27/%3e"
            style="display: block; max-width: 100%; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px;"
          />
        </span>
        <img
          alt="MN Logo"
          data-nimg="intrinsic"
          decoding="async"
          src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
          style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
        />
        <noscript />
      </span>
    </a>
  </main>
</div>
`;

exports[`PageTemplate successful request should render with empty routes array 1`] = `
<div>
  <main
    class="mx-4 my-4 p-2 pb-5 grow grid bg-base rel rounded"
  >
    <h1
      class="fs-2 text-center fg-primary leading-none mb-1 cols-12"
    >
      NexTrip: Metro Transit
    </h1>
    <h2
      class="fs-1.5 text-center mb-2 fg-secondary leading-none border-b cols-12 mx-auto"
    >
      Explore Minnesota
      <i
        class="block ml-.25 italic fs-1 fg-accent-primary"
      >
        Without Worrying About the Traffic
      </i>
    </h2>
    <div
      class="cols-12 flex flex-col mx-auto"
    >
      <label
        class="mb-.5 bold text-center"
        for="routes"
      >
        Choose a bus route:
      </label>
      <select
        class="fs-1 mb-1"
        id="routes"
        name="routes"
      >
        <option
          value=""
        >
          Select an available route
        </option>
      </select>
      <label
        class="mb-.5 bold text-center"
        for="directions"
      >
        Choose a direction:
      </label>
      <select
        class="fs-1 mb-1"
        disabled=""
        id="directions"
        name="directions"
      >
        <option
          value=""
        >
          Select a direction
        </option>
      </select>
    </div>
    <div
      class="cols-12"
    />
    <a
      aria-label="Learn more about MetroTransit NexTrip"
      class="pointer abs t-.25 r-.75"
      href="https://www.metrotransit.org/nextrip"
    >
      <span
        style="box-sizing: border-box; display: inline-block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: relative; max-width: 100%;"
      >
        <span
          style="box-sizing: border-box; display: block; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; max-width: 100%;"
        >
          <img
            alt=""
            aria-hidden="true"
            src="data:image/svg+xml,%3csvg%20xmlns=%27http://www.w3.org/2000/svg%27%20version=%271.1%27%20width=%2764%27%20height=%2732%27/%3e"
            style="display: block; max-width: 100%; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px;"
          />
        </span>
        <img
          alt="MN Logo"
          data-nimg="intrinsic"
          decoding="async"
          src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
          style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
        />
        <noscript />
      </span>
    </a>
  </main>
</div>
`;

exports[`fetcher should fetch and parse json 1`] = `
Array [
  Object {
    "agency_id": 0,
    "route_id": "901",
    "route_label": "METRO Blue Line",
  },
  Object {
    "agency_id": 0,
    "route_id": "902",
    "route_label": "METRO Green Line",
  },
  Object {
    "agency_id": 10,
    "route_id": "906",
    "route_label": "Airport Shuttle",
  },
  Object {
    "agency_id": 0,
    "route_id": "903",
    "route_label": "METRO Red Line",
  },
]
`;
